homeassistant/components/aurora_abb_powerone/__init__.py

```diff
@@ -45,3 +49,58 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
         hass.data[DOMAIN].pop(entry.entry_id)
 
     return unload_ok
+
+
+class AuroraAbbDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to manage fetching AuroraAbbPowerone data."""
+
+    def __init__(self, hass: HomeAssistant, comport: str, address: int) -> None:
+        """Initialize the data update coordinator."""
+        self.available_prev: bool = False
+        self.available: bool = False
+        self.client = AuroraSerialClient(address, comport, parity="N", timeout=1)
+        super().__init__(hass, _LOGGER, name=DOMAIN, update_interval=SCAN_INTERVAL)
+
+    async def _async_update_data(self) -> dict[str, float]:
+        """Fetch new state data for the sensor.
+
+        This is the only method that should fetch new data for Home Assistant.
+        """
+        data: dict[str, float] = {}
+        try:
+            self.available_prev = self.available
+            self.client.connect()
+
+            # read ADC channel 3 (grid power output)
+            power_watts = self.client.measure(3, True)
+            data["instantaneouspower"] = round(power_watts, 1)
+
+            temperature_c = self.client.measure(21)
+            data["temp"] = round(temperature_c, 1)
+
+            energy_wh = self.client.cumulated_energy(5)
+            data["totalenergy"] = round(energy_wh / 1000, 2)
+            self.available = True
+            self.data = data
+
+        except AuroraTimeoutError:
+            self.data = {}
+            self.available = False
+            _LOGGER.debug("No response from inverter (could be dark)")
+        except AuroraError as error:
+            self.data = {}
+            self.available = False
+            raise error
+        finally:
+            if self.available != self.available_prev:
+                if self.available:
+                    _LOGGER.info("Communication with %s back online", self.name)
+                else:
+                    _LOGGER.warning(
+                        "Communication with %s lost",
+                        self.name,
+                    )
+            if self.client.serline.isOpen():
+                self.client.close()
```

Comment:
Please test this line.  Should be just one more `patch` statement in an existing test.

Reply:
Thanks! Fixed in cfc384d.